File: dotnet/DesignPatterns/StructuralDesignPatterns/CompositeDesignPattern/CompositeDesignPattern.Example3/EmployeeComposite.cs
namespace CompositeDesignPattern.Example3;

public class EmployeeComposite : Employee
{
	protected List<Employee> _employees = new();
	protected EmployeeComposite(string name, double salary) : base(name, salary)
	{
	}

	public void AddEmployee(Employee employee)
		=> _employees.Add(employee);
	public void RemoveEmployee(Employee employee)
		=> _employees.Remove(employee);

	public override void Print()
	{
		Console.WriteLine($"Employee: {_name}, Salary: {_salary}");
		foreach (var employee in _employees)
		{
			employee.Print();
		}
	}
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/CompositeDesignPattern/CompositeDesignPattern.Example3/src/EmployeeComposite.ts
import { Employee } from './Employee';

export class EmployeeComposite extends Employee {
  protected employees: Employee[] = [];

  constructor(name: string, salary: number) { super(name, salary); }

  addEmployee(employee: Employee): void { this.employees.push(employee); }
  removeEmployee(employee: Employee): void {
    const idx = this.employees.indexOf(employee);
    if (idx !== -1) this.employees.splice(idx, 1);
  }

  print(): void {
    console.log(`Employee: ${this.name}, Salary: ${this.salary}`);
    for (const e of this.employees) e.print();
  }
}
